File: Exercises2024/Chapter04_basic_mc/Fortran/4-MCLennardJones/Source/mcmove.f
      Subroutine Mcmove(RunningEnergy, RunningVirial, Attempts, Accepted, 
     &                  MaxDisplacement)

      Implicit None

      Include 'parameter.inc'
      Include 'conf.inc'
      Include 'system.inc'

      Double Precision EnergyNew, EnergyOld, RunningEnergy, RandomNumber, 
     &     Xn, Yn, Zn, VirialOld, VirialNew, RunningVirial, MaxDisplacement
      Integer O, Attempts, Accepted
 
      Attempts = Attempts + 1

C     ---Select A Particle At Random

      O = Int(Dble(NumberOfParticles)*RandomNumber()) + 1

C     ---Calculate Energy Old Configuration

      Call EnergyParticle(X(O), Y(O), Z(O), O, 1, EnergyOld, VirialOld)

C     ---Give Particle A Random Displacement

      Xn = X(O) + (RandomNumber()-0.5d0)*MaxDisplacement
      Yn = Y(O) + (RandomNumber()-0.5d0)*MaxDisplacement
      Zn = Z(O) + (RandomNumber()-0.5d0)*MaxDisplacement

C     ---Calculate Energy New Configuration:

      Call EnergyParticle(Xn, Yn, Zn, O, 1, EnergyNew, VirialNew)

C     ---Acceptance Test

      If (RandomNumber().Lt.Exp(-Beta*(EnergyNew-EnergyOld))) Then

C     --Accepted

        Accepted = Accepted + 1
        RunningEnergy = RunningEnergy + (EnergyNew-EnergyOld)
        RunningVirial = RunningVirial + (VirialNew-VirialOld)
        
C        ---Put Particle In Simulation Box

        If (Xn.Lt.0.0d0) Then
          Xn = Xn + Box
        Elseif (Xn.Gt.Box) Then
          Xn = Xn - Box
        Endif
        
        If (Yn.Lt.0.0d0) Then
          Yn = Yn + Box
        Elseif (Yn.Gt.Box) Then
          Yn = Yn - Box
        Endif
        
        If (Zn.Lt.0.0d0) Then
          Zn = Zn + Box
        Elseif (Zn.Gt.Box) Then
          Zn = Zn - Box
        Endif
        
        X(O) = Xn
        Y(O) = Yn
        Z(O) = Zn
      End If
      Return
      End

File: Exercises2024/Chapter06_vle_free_energy/C/1-Widom/Source/mc_move.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "system.h"
#include "ran_uniform.h"

// attempts to displace a randomly selected particle
void Mcmove(void)
{
  double EnergyNew,VirialNew,EnergyOld,VirialOld;
  VECTOR NewPosition;
  int i;
 
  NumberOfAttempts++;

  // choose a random particle
  i=NumberOfParticles*RandomNumber();

  // calculate old energy
  EnergyParticle(Positions[i],i,0,&EnergyOld,&VirialOld);

  // give a random displacement
  NewPosition.x=Positions[i].x+(RandomNumber()-0.5)*MaximumDisplacement;
  NewPosition.y=Positions[i].y+(RandomNumber()-0.5)*MaximumDisplacement;
  NewPosition.z=Positions[i].z+(RandomNumber()-0.5)*MaximumDisplacement;

  // calculate new energy
  EnergyParticle(NewPosition,i,0,&EnergyNew,&VirialNew);
         
  if(RandomNumber()<exp(-Beta*(EnergyNew-EnergyOld)))
  {
    // accept
    NumberOfAcceptedMoves++;
    RunningEnergy+=(EnergyNew-EnergyOld);
    RunningVirial+=(VirialNew-VirialOld);

    // put particle in simulation box
    if(Positions[i].x<0.0)
      Positions[i].x+=Box;
    else if(Positions[i].x>=Box)
      Positions[i].x-=Box;
            
    if(Positions[i].y<0.0)
      Positions[i].y+=Box;
    else if(Positions[i].y>=Box)
      Positions[i].y-=Box;
          
    if(Positions[i].z<0.0)
      Positions[i].z+=Box;
    else if(Positions[i].z>=Box)
      Positions[i].z-=Box;

   // update new position
   Positions[i].x=NewPosition.x;
   Positions[i].y=NewPosition.y;
   Positions[i].z=NewPosition.z;
  }
}
